_posts/2024-03-20-kraft-migration.md

```diff
@@ -0,0 +1,196 @@
+For more details, please refer to the official [ZooKeeper to KRaft Migration](https://kafka.apache.org/documentation/#kraft_zk_migration) documentation.
+
+#### Deploying the KRaft controller quorum
+
+At the beginning, we have the Kafka brokers running in ZooKeeper-mode and connected to the ZooKeeper ensemble used to store metadata.
+
+![ZooKeeper-based cluster](/assets/images/posts/2024-03-20-kraft-migration-01-zk-brokers.png)
+
+> NOTE: the green square boxed number highlights the "generation" of the nodes which are rolled several times during the process.
+
+The first step is to deploy the KRaft controller quorum which will be in charge of maintaining the metadata in KRaft mode.
+In general, the number of KRaft controller nodes will be the same as the number of the ZooKeeper nodes currently running.
+It is also important to highlight that the migration process doesn't support migrating to nodes running in "combined" mode
+The nodes forming the KRaft controller quorum are all configured with the connection to ZooKeeper together with the additional `zookeeper.metadata.migration.enable=true` flag which states the intention to run the migration.
+When the KRaft controllers start, they form a quorum, elect the leader and move in a state where they are waiting for the brokers to register.
+
+![KRaft controller quorum deployed](/assets/images/posts/2024-03-20-kraft-migration-02-kraft-deployed.png)
+
+#### Enabling brokers to run the migration
+
+The next step is to move the brokers to migration mode.
+In order to do so, the brokers' configuration needs to be updated by adding connection details for all nodes of the KRaft controller quorum and enabling the migration with the `zookeeper.metadata.migration.enable=true` flag.
+After the update, the brokers need to be rolled one by one to make such configuration changes effective.
+On restart, the brokers register to the KRaft controller quorum and the migration begins.
+The KRaft controller leader copies all metadata from ZooKeeper to the `__cluster_metadata` topic.
+
+![KRaft migration running](/assets/images/posts/2024-03-20-kraft-migration-03-kraft-migration.png)
+
+While the migration is running, you can verify its status by looking at the log on the KRaft controller leader or by checking the `kafka.controller:type=KafkaController,name=ZkMigrationState` metric.
+When the migration is completed, that is the metric value is `MIGRATION`, the brokers are anyway still running in ZooKeeper mode.
+The KRaft controllers are in charge of handling any requests related to metadata changes within the cluster but they keep sending RPCs to the brokers for metadata updates.
+The metadata are still copied to ZooKeeper and the cluster is working in a so called "dual-write" mode.
+
+![KRaft dual-write](/assets/images/posts/2024-03-20-kraft-migration-04-kraft-dual-write.png)
+
+#### Moving brokers to be full KRaft
```

Comment:
This subheading is a bit weird. What about "Finalizing the broker migration" or "Restarting the brokers in KRaft mode"?

If you update this, also update the first sentence below to match